Cannot load tree details from API after setting diameter to invisible

Output of the Django logs:

```
HTTPError: HTTP Error 500: Internal Server Error
2014-08-14 13:41:15,246 [ERROR] django.request: Internal Server Error: /api/v3/instance/easton/locations/40.70203855347575,-75.20779062062502/plots
Traceback (most recent call last):
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 114, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/otm/env/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 57, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/otm/app/opentreemap/api/decorators.py", line 107, in wrapperf
    return view_f(request, *args, **kwargs)
  File "/usr/local/otm/app/opentreemap/api/decorators.py", line 151, in wrapper
    return view_f(request, *args, **kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 43, in wrapper
    return view_fn(request, instance, *args, **kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 208, in newreq
    outp = req_function(request, *args, **kwargs)
  File "/usr/local/otm/app/opentreemap/treemap/decorators.py", line 191, in newreq
    outp = req_function(request, *args, **kwargs)
  File "/usr/local/otm/app/opentreemap/api/plots.py", line 48, in plots_closest_to_point
    return [ctxt_for_plot(plot) for plot in plots]
  File "/usr/local/otm/app/opentreemap/api/plots.py", line 46, in ctxt_for_plot
    return context_dict_for_plot(request, plot)
  File "/usr/local/otm/app/opentreemap/treemap/lib/map_feature.py", line 159, in context_dict_for_plot
    context = context_dict_for_map_feature(request, plot)
  File "/usr/local/otm/app/opentreemap/treemap/lib/map_feature.py", line 304, in context_dict_for_map_feature
    _add_eco_benefits_to_context_dict(instance, feature, context)
  File "/usr/local/otm/app/opentreemap/treemap/lib/map_feature.py", line 72, in _add_eco_benefits_to_context_dict
    instance, feature)
  File "/usr/local/otm/app/opentreemap/treemap/ecobenefits.py", line 197, in benefits_for_object
    'eco.json', params.iteritems())
  File "/usr/local/otm/app/opentreemap/treemap/ecobackend.py", line 62, in json_benefits_call
    rslt = json.loads(urllib2.urlopen(req).read())
  File "/usr/lib/python2.7/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 406, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 519, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 444, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 378, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 527, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 500: Internal Server Error
```

--- a/opentreemap/treemap/ecobackend.py
+++ b/opentreemap/treemap/ecobackend.py
@@ -6,12 +6,47 @@ from __future__ import division
 import urllib2
 import urllib
 import json
+import re
 
 from django.conf import settings
 from django.contrib.gis.db.backends.postgis.adapter import PostGISAdapter
 from django.contrib.gis.geos import GEOSGeometry
 
-BAD_CODE_PAIR = 'bad code pair'
+import logging
+logger = logging.getLogger(__name__)
+
+
+# A system for handling unstructured text errors from the ecoservice
+#
+# Each key in `ECOBENEFIT_ERRORS` has a corresponding template
+# block in the plot_detail page that will display a user friendly
+# error message, as descriptive as possible.
+#
+# Each Value in `ECOBENEFIT_ERRORS` is a list of regexps used to
+# correctly identify a given error type. They will all be tried in
+# succession.
+#
+# If no pattern is found for any key, a 500 will raise.
+#
+# TODO: when the ecoservice starts returning structured error output
+# replace these patterns with simple destructuring.
+#
+
+BAD_CODE_PAIR = 'invalid_eco_pair'
+UNKNOWN_ECOBENEFIT_ERROR = 'unknown_ecobenefit_error'
+ECOBENEFIT_ERRORS = {
+    BAD_CODE_PAIR: [
+        (r'iTree code not found for otmcode '
+         r'([A-Z]+) in region ([A-Z][a-z]+[A-Z]+)\n')],
+    UNKNOWN_ECOBENEFIT_ERROR: [
+        'Species data not found for the .* region',
+        ('There are overrides defined for instance .* in '
+         'the .* region but not for species ID .*'),
+        ('There are overrides defined for the instance, '
+         'but not for the .* region'),
+        'Missing or invalid .* parameter',
+    ]
+}
 
 
 def json_benefits_call(endpoint, params, post=False, convert_params=True):
@@ -59,11 +94,14 @@ def json_benefits_call(endpoint, params, post=False, convert_params=True):
         req = url + '?' + paramString
 
     try:
-        rslt = json.loads(urllib2.urlopen(req).read())
+        return (json.loads(urllib2.urlopen(req).read()), None)
     except urllib2.HTTPError as e:
-        if e.fp.read() == 'invalid otm code for region\n':
-            return (None, BAD_CODE_PAIR)
+        error_body = e.fp.read()
+        logger.warning("ECOBENEFIT FAILURE: " + error_body)
+        for code, patterns in ECOBENEFIT_ERRORS.items():
+            for pattern in patterns:
+                match = re.match(pattern, error_body)
+                if match:
+                    return (None, code)
         else:
             raise
-
-    return (rslt, None)

--- a/opentreemap/treemap/lib/map_feature.py
+++ b/opentreemap/treemap/lib/map_feature.py
@@ -11,7 +11,7 @@ from django.utils.translation import ugettext as trans
 from django.db.models import Q
 
 from treemap.audit import Audit
-from treemap.ecobackend import BAD_CODE_PAIR
+from treemap.ecobackend import ECOBENEFIT_ERRORS
 from treemap.models import Tree, MapFeature, User
 
 from treemap.lib import format_benefits
@@ -72,8 +72,9 @@ def _add_eco_benefits_to_context_dict(instance, feature, context):
     benefits, basis, error = FeatureClass.benefits\
                                          .benefits_for_object(
                                              instance, feature)
-    if error == BAD_CODE_PAIR:
-        context['invalid_eco_pair'] = True
+
+    if error in ECOBENEFIT_ERRORS:
+        context[error] = True
     elif benefits:
         context.update(format_benefits(instance, benefits, basis))
 

--- a/opentreemap/treemap/templates/treemap/partials/plot_eco.html
+++ b/opentreemap/treemap/templates/treemap/partials/plot_eco.html
@@ -7,6 +7,12 @@
   <p>The tree's location falls within a climate region that does not support benefit calculations for this species.</p>
   {% endblocktrans %}
 </div>
+{% elif unknown_ecobenefit_error %}
+  {% blocktrans %}
+  <!-- TODO: When the ecoservice returns structured errors, replace this with real conditions -->
+  <!-- Or, maybe always leave this to avoid 500s -->
+  <p>{% trans "We were unable to calculate ecosystem benefits for this tree."}</p>
+  {% endblocktrans %}
 {% elif benefits.plot %}
 <table class="table table-hover">
   <tbody>

